src/_pytest/python.py

```diff
@@ -1144,19 +1144,17 @@ def _find_parametrized_scope(argnames, arg2fixturedefs, indirect):
 
 def _idval(val, argname, idx, idfn, item, config):
     if idfn:
-        s = None
         try:
-            s = idfn(val)
+            generated_id = idfn(val)
+            if generated_id is not None:
+                val = generated_id
         except Exception as e:
             # See issue https://github.com/pytest-dev/pytest/issues/2169
             msg = "{}: error raised while trying to determine id of parameter '{}' at position {}\n"
             msg = msg.format(item.nodeid, argname, idx)
             # we only append the exception type and message because on Python 2 reraise does nothing
             msg += "  {}: {}\n".format(type(e).__name__, e)
             six.raise_from(ValueError(msg), e)
-        if s:
-            return ascii_escaped(s)
```

Comment:
this is a behavior change, the hook now can override the given id
additionally i noted we font factor in the `ascii_escape` of the id from the hook

i propose factoring out the escaping logic to a function that gets a fallback result and using that to wrap around the return value in the different cases

Reply:
_headscratch_ I'm not seeing the behaviour change -- can you shed some light?  The only behaviour I intended to change is crash -> not crash (and restore what happened in pytest 3.0.7)

Reply:
by moving the idfn handling from returning before the `pytest_make_parametrize_id` to after invoking it,
the behavior is accidentally changed, as now the hook can override the result of the `idfn`, while before it couldn't

Reply:
That sounds like an improvement, no?

Reply:
as done here its possible that this inadvertely changes ids in a testsuite, so its a change for a major release

Reply:
it's easy to just `elif` I guess

Reply:
seems broken though that in this weird special case we don't apply the hook

Reply:
It is broken, its just that bugfixing is a breaking behavioral change

Reply:
how if fixing any bug not a behavioural change :rofl:

Reply:
@RonnyPfannschmidt better now?

Reply:
it seems much better, my gut is not 100% happy with the solution we have tho,
i'd really like to note down to change the behavior for pytest 6.0 (avoiding it in 5.0 for the management of the pytohn support drop)